openhtf/util/units/bin/units_from_xls.py

```diff
@@ -0,0 +1,160 @@
+
+python ./bin/units_from_xls.py --infile ./Rec20_Rev6e_2009.xls \
+--outfile ./__init__.py
+
+"""
+
+
+import argparse
+import os
+import shutil
+import sys
+from tempfile import mkstemp
+
+import xlrd
+
+
+# Column names for the columns we care about. This list must be populated in
+# the expected order: [<name label>, <code label>, <suffix label>].
+COLUMN_NAMES = ['Name',
+                'Common\nCode',
+                'Symbol']
+
+# The contents of the lines of the output file between which the generated code
+# will be inserted.
+START_DELIMITER = '######## BEGIN_UNITS ########'
+END_DELIMITER = '######## END_UNITS ########'
+
+SHEET_NAME = 'Annex I'
+REPLACEMENTS = {' ': '_',
+                '[': '',
+                ']': '',
+                '(': '',
+                ')': '',
+                ',': '.',
+                u'°': 'DEG_',
+               }
+
+
+def main():
+  """Main entry point for UNECE code .xls parsing."""
+  parser = argparse.ArgumentParser(description='Units From XLS',
+                                   prog='python units_from_xls.py')
+  parser.add_argument('--infile', type=str,
+                      help='A .xls file to parse.')
+  parser.add_argument('--outfile', type=str, default='__init__.py',
+                      help='Where to put the generated .py file.')
+  args = parser.parse_args()
+
+  code = code_generator(
+      xlrd.open_workbook(args.infile).sheet_by_name(SHEET_NAME))
+
+  insert_into_file(args.outfile, code, START_DELIMITER, END_DELIMITER)
+
+
+def code_generator(sheet):
+  """A generator that parses a worksheet containing UNECE code definitions.
+
+  Args:
+    sheet: An xldr.sheet object representing a UNECE code worksheet.
+  Yields: An iterable of lines of Python source code defining OpenHTF Units.
+  """
+  seen = set()
+  try:
+    col_indices = {}
+    rows = sheet.get_rows()
+    
+    # Find the indices for the columns we care about.
+    for idx, cell in enumerate(rows.next()):
+      if cell.value in COLUMN_NAMES:
+        col_indices[cell.value] = idx
+
+    # loop over all remaining rows and pull out units.
+    for row in rows:
+      name = row[col_indices[COLUMN_NAMES[0]]].value
+      code = row[col_indices[COLUMN_NAMES[1]]].value
+      suffix = multi_replace(row[col_indices[COLUMN_NAMES[2]]].value,
+                             {"'": "\\\'", '"': '\\\"'})
+      if name in seen:
+        continue
+      seen.add(name)
+
+      yield 'UNITS[\'%s\'] = Unit(\'%s\', \'%s\', \'%s\')\n' % (
+          multi_replace(name.upper(), REPLACEMENTS), name, code, suffix)
+
+  except xlrd.XLRDError:
+    sys.stdout.write('Unable to process the .xls file.')
+
+
+def multi_replace(string, replacements=None):
+  """Return the string in uppercase with the specified substitutions."""
+  replacements = replacements or {}
+  result = string
+
+  for old, new in replacements.iteritems():
+    result = result.replace(old, new)
+
+  return result
+
+
+def insert_into_file(filepath, code, start, end):
```

Comment:
This start/end stuff seems easy to break. If you just put the top and bottom of the file as a triple-quoted string here, then you could just blow away the old file each time.